File: GLLara/GLLModelMesh+OBJExport.m
//
//  GLLModelMesh+OBJExport.m
//  GLLara
//
//  Created by Torsten Kammer on 19.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLModelMesh+OBJExport.h"

#import "GLLVertexAttribAccessor.h"
#import "GLLVertexAttribAccessorSet.h"
#import "GLLVertexFormat.h"

#import "simd_matrix.h"

@implementation GLLModelMesh (OBJExport)

- (NSString *)writeOBJWithTransformations:(const mat_float16 *)transforms baseIndex:(uint32_t)baseIndex includeColors:(BOOL)includeColors;
{
    NSMutableString *objString = [[NSMutableString alloc] init];
    
    [objString appendFormat:@"g %@\n", [[self.name componentsSeparatedByCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] componentsJoinedByString:@"_"]];
    [objString appendFormat:@"usemtl material%lu\n", self.meshIndex];
    
    GLLVertexAttribAccessor *positionAccessor = [self.vertexDataAccessors accessorForSemantic:GLLVertexAttribPosition];
    GLLVertexAttribAccessor *normalAccessor = [self.vertexDataAccessors accessorForSemantic:GLLVertexAttribPosition];
    GLLVertexAttribAccessor *texCoordAccessor = [self.vertexDataAccessors accessorForSemantic:GLLVertexAttribTexCoord0 layer:0];
    GLLVertexAttribAccessor *colorAccessor = [self.vertexDataAccessors accessorForSemantic:GLLVertexAttribColor layer:0];
    GLLVertexAttribAccessor *boneIndexAccessor = [self.vertexDataAccessors accessorForSemantic:GLLVertexAttribBoneIndices];
    GLLVertexAttribAccessor *boneWeightAccessor = [self.vertexDataAccessors accessorForSemantic:GLLVertexAttribBoneWeights];
    
    for (NSInteger i = 0; i < self.countOfVertices; i++)
    {
        const float *position = [positionAccessor elementAt:i];
        const float *normal = [normalAccessor elementAt:i];
        
        mat_float16 transform = transforms[0];
        if (boneIndexAccessor)
        {
            const uint16_t *boneIndices = [boneIndexAccessor elementAt:i];
            const float *boneWeights = [boneWeightAccessor elementAt:i];
            
            transform = simd_mat_scale(transforms[boneIndices[0]], boneWeights[0]);
            transform = simd_mat_add(transform, simd_mat_scale(transforms[boneIndices[1]], boneWeights[1]));
            transform = simd_mat_add(transform, simd_mat_scale(transforms[boneIndices[2]], boneWeights[2]));
            transform = simd_mat_add(transform, simd_mat_scale(transforms[boneIndices[3]], boneWeights[3]));
        }
        
        vec_float4 transformedPosition = simd_mat_vecmul(transform, simd_make(position[0], position[1], position[2], 1.0f));
        vec_float4 transformedNormal = simd_mat_vecrotate(transform, simd_make(normal[0], normal[1], normal[2], 0.0f));
        
        [objString appendFormat:@"v %f %f %f\n", simd_extract(transformedPosition, 0), simd_extract(transformedPosition, 1), simd_extract(transformedPosition, 2)];
        [objString appendFormat:@"vn %f %f %f\n", simd_extract(transformedNormal, 0), simd_extract(transformedNormal, 1), simd_extract(transformedNormal, 2)];
        
        const float *texCoords = [texCoordAccessor elementAt:i];
        [objString appendFormat:@"vt %f %f\n", texCoords[0], 1.0 - texCoords[1]]; // Turn tex coords around (because I don't want to swap the whole image)
        
        if (includeColors)
        {
            const uint8_t *color = [colorAccessor elementAt:i];
            [objString appendFormat:@"vc %f %f %f %f\n", (float) color[0] / 255.0f, (float) color[1] / 255.0f, (float) color[2] / 255.0f, (float) color[3] / 255.0f];
        }
    }
    
    for (NSInteger i = 0; i < self.countOfUsedElements; i += 3)
    {
        NSUInteger adjustedElements[3] = {
            [self elementAt:i + 0] + baseIndex + 1,
            [self elementAt:i + 2] + baseIndex + 1,
            [self elementAt:i + 1] + baseIndex + 1
        };
        
        if (includeColors)
            [objString appendFormat:@"f %1$lu/%1$lu/%1$lu/%1$lu %2$lu/%2$lu/%2$lu/%2$lu %3$lu/%3$lu/%3$lu/%3$lu\n", adjustedElements[0], adjustedElements[1], adjustedElements[2]];
        else
            [objString appendFormat:@"f %1$lu/%1$lu/%1$lu %2$lu/%2$lu/%2$lu %3$lu/%3$lu/%3$lu\n", adjustedElements[0], adjustedElements[1], adjustedElements[2]];
    }
    
    return [objString copy];
}

@end

File: GLLara/GLLModelMesh+OBJExport.swift
//
//  GLLModelMesh+OBJExport.swift
//  GLLara
//
//  Created by Torsten Kammer on 06.04.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Foundation

extension GLLModelMesh {
    
    @objc func writeOBJ(transformations: UnsafePointer<mat_float16>, baseIndex: Int, includeColors: Bool) -> String {
        
        var objString = ""
        let groupName = name.components(separatedBy: CharacterSet.whitespacesAndNewlines).joined(separator: "_")
        objString.append("g \(groupName)\n")
        objString.append("usemtl material\(meshIndex)")
        
        let positionAccessor = vertexDataAccessors!.accessor(semantic: .position)!
        let normalAccessor = vertexDataAccessors!.accessor(semantic: .normal)!
        let texCoordAccessor = vertexDataAccessors!.accessor(semantic: .texCoord0, layer: 0)!
        let colorAccessor = vertexDataAccessors!.accessor(semantic: .color)!
        let boneIndexAccessor = vertexDataAccessors!.accessor(semantic: .boneIndices)
        let boneWeightAccessor = vertexDataAccessors!.accessor(semantic: .boneWeights)
        
        for i in 0..<countOfVertices {
            let position = positionAccessor.simd3Element(at: i, base: Float32.self)
            let normal = normalAccessor.simd3Element(at: i, base: Float.self)
            
            var transform = transformations[0]
            if let boneIndexAccessor = boneIndexAccessor, let boneWeightAccessor = boneWeightAccessor {
                let boneIndices = boneIndexAccessor.typedElementArray(at: i, type: UInt16.self)
                let boneWeights = boneWeightAccessor.typedElementArray(at: i, type: Float.self)
                
                transform = simd_linear_combination(boneWeights[0], transformations[Int(boneIndices[0])], boneWeights[1], transformations[Int(boneIndices[1])]) + simd_linear_combination(boneWeights[2], transformations[Int(boneIndices[2])], boneWeights[3], transformations[Int(boneIndices[3])])
            }
            
            let transformedPosition = simd_mul(transform, vec_float4(position, 1.0))
            let transformedNormal = simd_mat_vecrotate(transform, vec_float4(normal, 0.0))
            
            objString.append("v \(transformedPosition.x) \(transformedPosition.y) \(transformedPosition.z)")
            
            objString.append("vn \(transformedNormal.x) \(transformedNormal.y) \(transformedNormal.z)")
            
            let texCoords = texCoordAccessor.typedElementArray(at: i, type: Float.self)
            objString.append("vt \(texCoords[0]) \(1.0 - texCoords[1])") // Turn tex coords around (because I don't want to swap the whole image)
            
            if includeColors {
                let color = colorAccessor.typedElementArray(at: i, type: UInt8.self)
                let r = Double(color[0]) / 255.0
                let g = Double(color[1]) / 255.0
                let b = Double(color[2]) / 255.0
                let a = Double(color[3]) / 255.0
                objString.append("vc \(r) \(g) \(b) \(a)")
            }
        }
        
        for i in stride(from: 0, to: countOfUsedElements, by: 3) {
            let adjustedElements = [
                element(at: i + 0) + baseIndex + 1,
                element(at: i + 2) + baseIndex + 1,
                element(at: i + 1) + baseIndex + 1
            ]
            objString.append("f")
            if includeColors {
                for element in adjustedElements {
                    objString.append(" \(element)/\(element)/\(element)/\(element)")
                }
            } else {
                for element in adjustedElements {
                    objString.append(" \(element)/\(element)/\(element)")
                }
            }
            objString.append("\n")
        }
        
        return objString
    }
    
}
